src/hardware/input/mouseif_virtual_machines.cpp

```diff
@@ -62,124 +48,199 @@ union VMwareButtons {
-	                                                 : 0x100 + counter_w);
+	if (is_absolute) {
+		mouse_vmm_activate(VmmProtocol::VirtualBox);
+	} else {
+		mouse_vmm_deactivate(VmmProtocol::VirtualBox);
+	}
+}
 
-	counter_w = 0;
+void MOUSE_VirtualBox_SetPointerVisible(const bool is_visible)
+{
+	if (virtualbox.wants_pointer != is_visible) {
+		virtualbox.wants_pointer = is_visible;
+		if (virtualbox.is_active) {
+			mouse_shared.vmm_wants_pointer = is_visible;
+			MOUSE_UpdateGFX();
+		}
+	}
 }
 
-static void cmd_abs_pointer_status()
+void MOUSE_VirtualBox_GetPointerStatus(VirtualBoxPointerStatus& status)
 {
-	reg_eax = updated ? ABS_UPDATED : ABS_NOT_UPDATED;
-	updated = false;
+	status.absolute_x = scaled_x;
+	status.absolute_y = scaled_y;
 }
 
-static void cmd_abs_pointer_command()
+// ***************************************************************************
+// VMware interface implementation
+// ***************************************************************************
+
+bool MOUSE_VMware_IsSupported()
 {
-	switch (static_cast<VMwareAbsPointer>(reg_ebx)) {
-	case VMwareAbsPointer::Enable: break; // can be safely ignored
-	case VMwareAbsPointer::Relative: MOUSEVMM_Deactivate(); break;
-	case VMwareAbsPointer::Absolute: MOUSEVMM_Activate(); break;
-	default:
-		LOG_WARNING("MOUSE (PS/2): unimplemented VMware subcommand 0x%08x",
-		            reg_ebx);
-		break;
-	}
+	return mouse_config.vmware_mouse &&
+	       (mouse_config.model_ps2 != MouseModelPS2::NoMouse);
 }
 
-static uint32_t port_read_vmware(const io_port_t, const io_width_t)
+void MOUSE_VMware_Activate()
 {
-	if (reg_eax != VMWARE_MAGIC)
-		return 0;
-
-	switch (static_cast<VMwareCmd>(reg_cx)) {
-	case VMwareCmd::GetVersion: cmd_get_version(); break;
-	case VMwareCmd::AbsPointerData: cmd_abs_pointer_data(); break;
-	case VMwareCmd::AbsPointerStatus: cmd_abs_pointer_status(); break;
-	case VMwareCmd::AbsPointerCommand: cmd_abs_pointer_command(); break;
-	default:
-		LOG_WARNING("MOUSE (PS/2): unimplemented VMware command 0x%08x",
-		            reg_ecx);
-		break;
-	}
+	mouse_vmm_activate(VmmProtocol::VMware);
+}
+
+void MOUSE_VMware_Deactivate()
+{
+	mouse_vmm_deactivate(VmmProtocol::VMware);
+}
+
+bool MOUSE_VMware_CheckIfUpdated()
+{
+	const bool result = vmware.updated;
+	vmware.updated    = false;
+
+	return result;
+}
+
+void MOUSE_VMware_GetPointerStatus(VMwarePointerStatus& status)
+{
+	status.absolute_x = scaled_x;
+	status.absolute_y = scaled_y;
+
+	status.buttons       = vmware.buttons._data;
+	status.wheel_counter = static_cast<uint8_t>(vmware.counter_w);
 
-	return reg_eax;
+	vmware.counter_w = 0;
+}
+
+// ***************************************************************************
+// External notifications
+// ***************************************************************************
+
+void MOUSEVMM_Deactivate()
+{
+	mouse_vmm_deactivate(VmmProtocol::VirtualBox);
+	mouse_vmm_deactivate(VmmProtocol::VMware);
```

Comment:
Minor thing, but you're kinda using `vmm` and `VMware` interchangeably? What's `vmm`? VMware Mouse or something?

Reply:
_VMM_ stands for _Virtual Machine Manager_ (https://en.wikipedia.org/wiki/Hypervisor) and is used in function/variable names common to _VMware_ and _VirtualBox_. Shall I change it? Maybe to `virtualizer`?

Reply:
Yeah, I think VMM is fine, just can you explain what it stands for at top of the file or somewhere?

Reply:
OK, added explanations in several places